# WINE DATASET - PCA

    Athul Mathew  CB.EN.P2AID20016   PCA

Dataset URL: https://github.com/toshihiroryuu/Dataset-test/blob/main/Wine.csv

In [1]:
import pandas as pd

url = "https://raw.githubusercontent.com/toshihiroryuu/Dataset-test/main/Wine.csv"
# read wine data set. 
df = pd.read_csv(url)

df.head()

Type  Alcohol  Malic_Acid   Ash  Ash_Alcalinity  Magnesium  Total_Phenols  \
0    A    14.23        1.71  2.43            15.6        127           2.80   
1    A    13.20        1.78  2.14            11.2        100           2.65   
2    A    13.16        2.36  2.67            18.6        101           2.80   
3    A    14.37        1.95  2.50            16.8        113           3.85   
4    A    13.24        2.59  2.87            21.0        118           2.80   

   Flavanoids  Nonflavanoid_Phenols  Proanthocyanins  Color_Intensity   Hue  \
0        3.06                  0.28             2.29             5.64  1.04   
1        2.76                  0.26             1.28             4.38  1.05   
2        3.24                  0.30             2.81             5.68  1.03   
3        3.49                  0.24             2.18             7.80  0.86   
4        2.69                  0.39             1.82             4.32  1.04   

   OD280_OD315  Proline  
0         3.92     1065  
1         3.40     1050  
2         3.17     1185  
3         3.45     1480  
4         2.93      735

In [2]:
# identify the size of data frame
print("Size of the data frame is",df.size)

# identify the shape of the data frame 
print("Shape of the data frame is",df.shape)
print("No of Rows = ", df.shape[0])
print("No of Columns = ", df.shape[1])

df.describe()

Size of the data frame is 2492
Shape of the data frame is (178, 14)
No of Rows =  178
No of Columns =  14


Alcohol  Malic_Acid         Ash  Ash_Alcalinity   Magnesium  \
count  178.000000  178.000000  178.000000      178.000000  178.000000   
mean    13.000618    2.336348    2.366517       19.494944   99.741573   
std      0.811827    1.117146    0.274344        3.339564   14.282484   
min     11.030000    0.740000    1.360000       10.600000   70.000000   
25%     12.362500    1.602500    2.210000       17.200000   88.000000   
50%     13.050000    1.865000    2.360000       19.500000   98.000000   
75%     13.677500    3.082500    2.557500       21.500000  107.000000   
max     14.830000    5.800000    3.230000       30.000000  162.000000   

       Total_Phenols  Flavanoids  Nonflavanoid_Phenols  Proanthocyanins  \
count     178.000000  178.000000            178.000000       178.000000   
mean        2.295112    2.029270              0.361854         1.590899   
std         0.625851    0.998859              0.124453         0.572359   
min         0.980000    0.340000              0.130000         0.410000   
25%         1.742500    1.205000              0.270000         1.250000   
50%         2.355000    2.135000              0.340000         1.555000   
75%         2.800000    2.875000              0.437500         1.950000   
max         3.880000    5.080000              0.660000         3.580000   

       Color_Intensity         Hue  OD280_OD315      Proline  
count       178.000000  178.000000   178.000000   178.000000  
mean          5.058090    0.957449     2.611685   746.893258  
std           2.318286    0.228572     0.709990   314.907474  
min           1.280000    0.480000     1.270000   278.000000  
25%           3.220000    0.782500     1.937500   500.500000  
50%           4.690000    0.965000     2.780000   673.500000  
75%           6.200000    1.120000     3.170000   985.000000  
max          13.000000    1.710000     4.000000  1680.000000

In [3]:
pd.DataFrame({'mean': df.mean(),
'sd': df.std(),
'min': df.min(),
'max': df.max(),
'median': df.median(),
'length': len(df),
'miss.val': df.isnull().sum(),
})

mean          sd    min    max   median  length  \
Alcohol                13.000618    0.811827  11.03  14.83   13.050     178   
Ash                     2.366517    0.274344   1.36   3.23    2.360     178   
Ash_Alcalinity         19.494944    3.339564   10.6     30   19.500     178   
Color_Intensity         5.058090    2.318286   1.28     13    4.690     178   
Flavanoids              2.029270    0.998859   0.34   5.08    2.135     178   
Hue                     0.957449    0.228572   0.48   1.71    0.965     178   
Magnesium              99.741573   14.282484     70    162   98.000     178   
Malic_Acid              2.336348    1.117146   0.74    5.8    1.865     178   
Nonflavanoid_Phenols    0.361854    0.124453   0.13   0.66    0.340     178   
OD280_OD315             2.611685    0.709990   1.27      4    2.780     178   
Proanthocyanins         1.590899    0.572359   0.41   3.58    1.555     178   
Proline               746.893258  314.907474    278   1680  673.500     178   
Total_Phenols           2.295112    0.625851   0.98   3.88    2.355     178   
Type                         NaN         NaN      A      C      NaN     178   

                      miss.val  
Alcohol                      0  
Ash                          0  
Ash_Alcalinity               0  
Color_Intensity              0  
Flavanoids                   0  
Hue                          0  
Magnesium                    0  
Malic_Acid                   0  
Nonflavanoid_Phenols         0  
OD280_OD315                  0  
Proanthocyanins              0  
Proline                      0  
Total_Phenols                0  
Type                         0

In [4]:
df = df.drop(['Type'], axis=1)
df.head(5)

Alcohol  Malic_Acid   Ash  Ash_Alcalinity  Magnesium  Total_Phenols  \
0    14.23        1.71  2.43            15.6        127           2.80   
1    13.20        1.78  2.14            11.2        100           2.65   
2    13.16        2.36  2.67            18.6        101           2.80   
3    14.37        1.95  2.50            16.8        113           3.85   
4    13.24        2.59  2.87            21.0        118           2.80   

   Flavanoids  Nonflavanoid_Phenols  Proanthocyanins  Color_Intensity   Hue  \
0        3.06                  0.28             2.29             5.64  1.04   
1        2.76                  0.26             1.28             4.38  1.05   
2        3.24                  0.30             2.81             5.68  1.03   
3        3.49                  0.24             2.18             7.80  0.86   
4        2.69                  0.39             1.82             4.32  1.04   

   OD280_OD315  Proline  
0         3.92     1065  
1         3.40     1050  
2         3.17     1185  
3         3.45     1480  
4         2.93      735

### Handle missing values

In [5]:
miss_count = df.isnull().sum().sum()

print("No of missing values is",miss_count)

No of missing values is 0


No missing values in the dataset.

### Normalise data

In [6]:
from sklearn import preprocessing

np_norm = preprocessing.scale(df)
df_norm = pd.DataFrame(np_norm, columns = df.columns)

print("Shape of the data frame is",df_norm.shape)
df_norm.head(5)

Shape of the data frame is (178, 13)


Alcohol  Malic_Acid       Ash  Ash_Alcalinity  Magnesium  Total_Phenols  \
0  1.518613   -0.562250  0.232053       -1.169593   1.913905       0.808997   
1  0.246290   -0.499413 -0.827996       -2.490847   0.018145       0.568648   
2  0.196879    0.021231  1.109334       -0.268738   0.088358       0.808997   
3  1.691550   -0.346811  0.487926       -0.809251   0.930918       2.491446   
4  0.295700    0.227694  1.840403        0.451946   1.281985       0.808997   

   Flavanoids  Nonflavanoid_Phenols  Proanthocyanins  Color_Intensity  \
0    1.034819             -0.659563         1.224884         0.251717   
1    0.733629             -0.820719        -0.544721        -0.293321   
2    1.215533             -0.498407         2.135968         0.269020   
3    1.466525             -0.981875         1.032155         1.186068   
4    0.663351              0.226796         0.401404        -0.319276   

        Hue  OD280_OD315   Proline  
0  0.362177     1.847920  1.013009  
1  0.406051     1.113449  0.965242  
2  0.318304     0.788587  1.395148  
3 -0.427544     1.184071  2.334574  
4  0.362177     0.449601 -0.037874

### PCA

In [7]:
from sklearn.decomposition import PCA
from sklearn import preprocessing

pcs = PCA()
pcs.fit(df_norm)

PCA()

In [8]:
import numpy as np

pcsSummary_df = pd.DataFrame({"Std devn":np.sqrt(pcs.explained_variance_),
                              "Proportion of var": pcs.explained_variance_ratio_,
                              "Cum proportion": np.cumsum(pcs.explained_variance_ratio_)})

In [9]:
pcsSummary_df = pcsSummary_df.transpose()
pcsSummary_df.columns = ["PC"+str(i) for i in range(1,len(pcsSummary_df.columns) + 1)]

In [10]:
pcsSummary_df.round(4)

PC1     PC2     PC3     PC4     PC5     PC6     PC7  \
Std devn           2.1754  1.5846  1.2059  0.9613  0.9263  0.8033  0.7444   
Proportion of var  0.3620  0.1921  0.1112  0.0707  0.0656  0.0494  0.0424   
Cum proportion     0.3620  0.5541  0.6653  0.7360  0.8016  0.8510  0.8934   

                      PC8     PC9    PC10    PC11   PC12    PC13  
Std devn           0.5920  0.5390  0.5023  0.4765  0.412  0.3224  
Proportion of var  0.0268  0.0222  0.0193  0.0174  0.013  0.0080  
Cum proportion     0.9202  0.9424  0.9617  0.9791  0.992  1.0000

    Principal Components 1 and 2 (PC1 & PC2) contains 55% of the information cumulatively.Principal Components 1 and 2 (PC1 & PC2) contains and 36% and 19.2 % of the information individually.

In [11]:
pcsComponents_df = pd.DataFrame(pcs.components_.transpose(),
                                columns=pcsSummary_df.columns,
                                index=df.columns)

pcsComponents_df

PC1       PC2       PC3       PC4       PC5  \
Alcohol               0.144329 -0.483652 -0.207383 -0.017856 -0.265664   
Malic_Acid           -0.245188 -0.224931  0.089013  0.536890  0.035214   
Ash                  -0.002051 -0.316069  0.626224 -0.214176 -0.143025   
Ash_Alcalinity       -0.239320  0.010591  0.612080  0.060859  0.066103   
Magnesium             0.141992 -0.299634  0.130757 -0.351797  0.727049   
Total_Phenols         0.394661 -0.065040  0.146179  0.198068 -0.149318   
Flavanoids            0.422934  0.003360  0.150682  0.152295 -0.109026   
Nonflavanoid_Phenols -0.298533 -0.028779  0.170368 -0.203301 -0.500703   
Proanthocyanins       0.313429 -0.039302  0.149454  0.399057  0.136860   
Color_Intensity      -0.088617 -0.529996 -0.137306  0.065926 -0.076437   
Hue                   0.296715  0.279235  0.085222 -0.427771 -0.173615   
OD280_OD315           0.376167  0.164496  0.166005  0.184121 -0.101161   
Proline               0.286752 -0.364903 -0.126746 -0.232071 -0.157869   

                           PC6       PC7       PC8       PC9      PC10  \
Alcohol              -0.213539 -0.056396 -0.396139  0.508619  0.211605   
Malic_Acid           -0.536814  0.420524 -0.065827 -0.075283 -0.309080   
Ash                  -0.154475 -0.149171  0.170260 -0.307694 -0.027125   
Ash_Alcalinity        0.100825 -0.286969 -0.427970  0.200449  0.052799   
Magnesium            -0.038144  0.322883  0.156361  0.271403  0.067870   
Total_Phenols         0.084122 -0.027925  0.405934  0.286035 -0.320131   
Flavanoids            0.018920 -0.060685  0.187245  0.049578 -0.163151   
Nonflavanoid_Phenols  0.258594  0.595447  0.233285  0.195501  0.215535   
Proanthocyanins       0.533795  0.372139 -0.368227 -0.209145  0.134184   
Color_Intensity       0.418644 -0.227712  0.033797  0.056218 -0.290775   
Hue                  -0.105983  0.232076 -0.436624  0.085828 -0.522399   
OD280_OD315          -0.265851 -0.044764  0.078108  0.137227  0.523706   
Proline              -0.119726  0.076805 -0.120023 -0.575786  0.162116   

                          PC11      PC12      PC13  
Alcohol              -0.225917 -0.266286  0.014970  
Malic_Acid            0.076486  0.121696  0.025964  
Ash                  -0.498691 -0.049622 -0.141218  
Ash_Alcalinity        0.479314 -0.055743  0.091683  
Magnesium             0.071289  0.062220  0.056774  
Total_Phenols         0.304341 -0.303882 -0.463908  
Flavanoids           -0.025694 -0.042899  0.832257  
Nonflavanoid_Phenols  0.116896  0.042352  0.114040  
Proanthocyanins      -0.237363 -0.095553 -0.116917  
Color_Intensity       0.031839  0.604222 -0.011993  
Hue                  -0.048212  0.259214 -0.089889  
OD280_OD315           0.046423  0.600959 -0.156718  
Proline               0.539270 -0.079402  0.014447

#### Weights of PC1.

In [12]:
pcs.components_[0]

array([ 0.1443294 , -0.24518758, -0.00205106, -0.23932041,  0.14199204,
        0.39466085,  0.4229343 , -0.2985331 ,  0.31342949, -0.0886167 ,
        0.29671456,  0.37616741,  0.28675223])

In [13]:
pcsComponents_df["PC1"]

Alcohol                 0.144329
Malic_Acid             -0.245188
Ash                    -0.002051
Ash_Alcalinity         -0.239320
Magnesium               0.141992
Total_Phenols           0.394661
Flavanoids              0.422934
Nonflavanoid_Phenols   -0.298533
Proanthocyanins         0.313429
Color_Intensity        -0.088617
Hue                     0.296715
OD280_OD315             0.376167
Proline                 0.286752
Name: PC1, dtype: float64

### Analysis

In [14]:
pcsSummary_df.round(4)

PC1     PC2     PC3     PC4     PC5     PC6     PC7  \
Std devn           2.1754  1.5846  1.2059  0.9613  0.9263  0.8033  0.7444   
Proportion of var  0.3620  0.1921  0.1112  0.0707  0.0656  0.0494  0.0424   
Cum proportion     0.3620  0.5541  0.6653  0.7360  0.8016  0.8510  0.8934   

                      PC8     PC9    PC10    PC11   PC12    PC13  
Std devn           0.5920  0.5390  0.5023  0.4765  0.412  0.3224  
Proportion of var  0.0268  0.0222  0.0193  0.0174  0.013  0.0080  
Cum proportion     0.9202  0.9424  0.9617  0.9791  0.992  1.0000

We have 13 Principal Componeents(PC's).
                                PC1 to PC9 holds around 95% of the information.
                                PC1 has the highest Standard deviation and Propotion of variance.
                                PC13 has the lowest Propotion of variance.
                                